modules/lib/lib-event/src/main/resources/lib/xp/event.ts

```diff
@@ -0,0 +1,136 @@
+import * as Lib from './event';
+
+/*
+ * Global library definitions for require and __non_webpack_import__
+ */
+
+type LibType = typeof Lib;
+
+declare global {
+    interface XpEventLibrary extends LibType {}
+}
+
+
+export interface ListenerParams<EventData extends object = EnonicEventData> {
+    /**
+     * Event type pattern
+     */
+    type: EnonicEventTypes;
+
+    /**
+     * Callback event listener
+     */
+    callback: (event: EnonicEvent<EventData>) => void;
+
+    /**
+     * Local events only (default to false)
+     */
+    localOnly?: boolean;
+}
+
+export interface SendParams {
+    /**
+     * Event type
+     */
+    type: EnonicEventTypes;
+
+    /**
+     * `true` if it should be distributed in cluster
+     */
+    distributed?: boolean;
+
+    /**
+     * Additional data for event.
+     */
+    data?: object;
+}
+
+export interface EnonicEvent<EventData extends object = EnonicEventData> {
+    readonly type: EnonicEventTypes | string;
+    readonly timestamp: number;
+    readonly localOrigin: boolean;
+    readonly distributed: boolean;
+    readonly data: EventData;
+}
+
+export interface EnonicEventData {
+    readonly nodes: ReadonlyArray<EnonicEventDataNode>;
+    readonly state?: string; // event type <'node.stateUpdated'>
+}
+
+export interface EnonicEventDataNode {
+    readonly id: string;
+    readonly path: string;
+    readonly branch: string;
+    readonly repo: string;
+    readonly newPath?: string; // event type <'node.moved' | 'node.renamed'>
+}
+
+type LiteralUnion<T extends U, U = string> = T | (U & Record<never, never>);
+
+export type EnonicEventTypes =
+    LiteralUnion<
+    | 'node.*'
+    | 'node.created'
+    | 'node.deleted'
+    | 'node.pushed'
+    | 'node.duplicated'
+    | 'node.updated'
+    | 'node.moved'
+    | 'node.renamed'
+    | 'node.sorted'
+    | 'node.stateUpdated'
+    | 'node.permissionsUpdated'
+    >;
+
+/**
+ * Event functions.
+ *
+ * @example
+ * var eventLib = require('/lib/xp/event');
+ *
+ * @module event
+ */
+
+/**
+ * This function adds a event listener.
+ *
+ * @example-ref examples/event/listener.js
+ *
+ * @param {object} params Listener parameters.
+ * @param {string} params.type Event type pattern.
+ * @param {function} params.callback Callback event listener.
+ * @param {boolean} params.localOnly Local events only (default to false).
+ */
+export function listener<EventData extends object = EnonicEventData>(params: ListenerParams<EventData>): void {
+    const helper = __.newBean<any>('com.enonic.xp.lib.event.EventListenerHelper');
```

Comment:
I think you should create interfaces to use with `__.newBean()`, because this is where the boundary between TypeScript and Java is. And therefore maybe the most "dangerous" place in the code. 

This might save you later (specially when it comes to nullability).